sdk/contracts-provider-service/src/implementation/contracts/Erc20Contract/Erc20Contract.ts

```diff
@@ -0,0 +1,104 @@
+import { IBlockchainClient } from '@summerfi/blockchain-client-provider'
+import { IErc20Contract } from '@summerfi/contracts-provider-common'
+import {
+  IAddress,
+  IChainInfo,
+  IToken,
+  ITokenAmount,
+  Maybe,
+  Token,
+  TokenAmount,
+  TransactionInfo,
+} from '@summerfi/sdk-common'
+import { Abi, encodeFunctionData, erc20Abi } from 'viem'
+import { ContractWrapper } from '../ContractWrapper'
+
+/**
+ * @name Erc20Contract
+ * @description Implementation for the ERC20 contract wrapper
+ * @implements IErc20Contract
+ */
+export class Erc20Contract<const TClient extends IBlockchainClient, TAddress extends IAddress>
+  extends ContractWrapper<typeof erc20Abi, TClient, TAddress>
+  implements IErc20Contract
+{
+  private _token: Maybe<IToken>
+
+  /** CONSTRUCTOR */
+  constructor(params: { blockchainClient: TClient; chainInfo: IChainInfo; address: TAddress }) {
+    super(params)
+  }
+
+  /** PUBLIC */
+
+  /** READ METHODS */
+
+  /** @see IErc20Contract.getToken */
+  async getToken(): Promise<IToken> {
+    if (!this._token) {
+      this._token = await this._retrieveTokenInfo()
+    }
+    return this._token
+  }
+
+  /** @see IErc20Contract.balanceOf */
+  async balanceOf(params: { address: IAddress }): Promise<ITokenAmount> {
```

Comment:
I think this is misleading, I just got caught by it in tests. usually balance is returned in base unit, but this is token amount which store value as not base unit. I would propose rename method to tokenAmountOf